services/src/pro/contexts/db_types.rs

```diff
@@ -0,0 +1,236 @@
+pub struct StacApiRetriesDbType {
+    pub number_of_retries: i64,
+    pub initial_delay_ms: i64,
+    pub exponential_backoff_factor: f64,
+}
+
+impl From<&StacApiRetries> for StacApiRetriesDbType {
+    fn from(other: &StacApiRetries) -> Self {
+        Self {
+            number_of_retries: other.number_of_retries as i64,
+            initial_delay_ms: other.initial_delay_ms as i64,
+            exponential_backoff_factor: other.exponential_backoff_factor,
+        }
+    }
+}
+
+impl TryFrom<StacApiRetriesDbType> for StacApiRetries {
+    type Error = Error;
+
+    fn try_from(other: StacApiRetriesDbType) -> Result<Self, Self::Error> {
+        Ok(Self {
+            number_of_retries: other.number_of_retries as usize,
+            initial_delay_ms: other.initial_delay_ms as u64,
+            exponential_backoff_factor: other.exponential_backoff_factor,
+        })
+    }
+}
+
+#[derive(Debug, ToSql, FromSql)]
+#[postgres(name = "GdalRetries")]
+pub struct GdalRetriesDbType {
+    pub number_of_retries: i64,
+}
+
+impl From<&GdalRetries> for GdalRetriesDbType {
+    fn from(other: &GdalRetries) -> Self {
+        Self {
+            number_of_retries: other.number_of_retries as i64,
+        }
+    }
+}
+
+impl TryFrom<GdalRetriesDbType> for GdalRetries {
+    type Error = Error;
+
+    fn try_from(other: GdalRetriesDbType) -> Result<Self, Self::Error> {
+        Ok(Self {
+            number_of_retries: other.number_of_retries as usize,
+        })
+    }
+}
+
+#[derive(Debug, ToSql, FromSql)]
+#[postgres(name = "ProDataProviderDefinition")]
+pub struct TypedProDataProviderDefinitionDbType {
+    sentinel_s2_l2_a_cogs_provider_definition: Option<SentinelS2L2ACogsProviderDefinition>,
+}
+
+impl From<&TypedProDataProviderDefinition> for TypedProDataProviderDefinitionDbType {
+    fn from(other: &TypedProDataProviderDefinition) -> Self {
+        match other {
+            TypedProDataProviderDefinition::SentinelS2L2ACogsProviderDefinition(
+                data_provider_definition,
+            ) => Self {
+                sentinel_s2_l2_a_cogs_provider_definition: Some(data_provider_definition.clone()),
+            },
+        }
+    }
+}
+
+impl TryFrom<TypedProDataProviderDefinitionDbType> for TypedProDataProviderDefinition {
+    type Error = Error;
+
+    #[allow(clippy::too_many_lines)]
+    fn try_from(
+        result_descriptor: TypedProDataProviderDefinitionDbType,
+    ) -> Result<Self, Self::Error> {
+        match result_descriptor {
+            TypedProDataProviderDefinitionDbType {
+                sentinel_s2_l2_a_cogs_provider_definition: Some(data_provider_definition),
+            } => Ok(
+                TypedProDataProviderDefinition::SentinelS2L2ACogsProviderDefinition(
+                    data_provider_definition,
+                ),
+            ),
+
+            _ => Err(Error::UnexpectedInvalidDbTypeConversion),
+        }
+    }
+}
+
+delegate_from_to_sql!(GdalRetries, GdalRetriesDbType);
```

Comment:
why can we not implement ToSql for GdalRetries?

Reply:
We could, but I strived for simplicity with these conversion types. Aren't the internals copy?